ctx.go

```diff
@@ -52,20 +45,22 @@ func PushNotify(ctx context.Context, method string, params interface{}) error {
 // A successful callback reports a nil error and a non-nil response. Errors
 // returned by the client have concrete type *jrpc2.Error.
 func PushCall(ctx context.Context, method string, params interface{}) (*Response, error) {
-	s := ctx.Value(serverKey{}).(*Server)
-	if !s.allowP {
-		return nil, ErrPushUnsupported
-	}
-	return s.Callback(ctx, method, params)
+	return ServerFromContext(ctx).Callback(ctx, method, params)
 }
 
 // CancelRequest requests the server associated with ctx to cancel the pending
 // or in-flight request with the specified ID.  If no request exists with that
 // ID, this is a no-op without error.
-func CancelRequest(ctx context.Context, id string) {
-	s := ctx.Value(serverKey{}).(*Server)
-	s.cancelRequests(ctx, []json.RawMessage{json.RawMessage(id)})
-}
+func CancelRequest(ctx context.Context, id string) { ServerFromContext(ctx).CancelRequest(id) }
+
+// ServerFromContext returns the server associated with the given context.
+// This will be populated on the context passed to request handlers.
+//
+// It is safe to retain the server and invoke its methods beyond the lifetime
+// of the context from which it was extracted; however, a handler must not
+// block on the Wait or WaitStatus methods of the server, as the server will
+// deadlock waiting for the handler to return.
+func ServerFromContext(ctx context.Context) *Server { return ctx.Value(serverKey{}).(*Server) }
```

Comment:
I know this makes the other methods a little more verbose, but is there a reason we couldn't do safe casting here to avoid panics? i.e. return `bool` as second value to indicate whether the context actually does contain `*Server`?

Of course when called from within a handler this should never happen, but I feel this function is poised to be accidentally called from other places with a wrong context.

Reply:
A program that calls this function outside a handler is already statically wrong—though in a way the type checker can't detect. I'd prefer it panic in that case—and chose the single-position assertion for that reason. I definitely don't want it exposed in the API, either way—though maybe you have a case for it returning nil?

Reply:
That's fine. After giving it a second thought I admit you're right - panic is actually probably better option in this case. It is something I'd _want_ to know about.

Reply:
I updated the comments on these helpers to make that more explicit.